File: mirror/tier2-php/showcase/sec_injection.php
<?php
// MIRROR V3: Tier 2 PHP Showcase
// Category: Security
// Module: sec_injection.php

function escape_sql(string $value): string {
    return str_replace(
        ["'", '"', "\\"],
        ["''", '""', "\\\\"],
        $value
    );
}

function validate_query(string $query): bool {
    $forbidden = ['DROP', 'DELETE', 'UPDATE', 'INSERT', 'ALTER'];
    $upperQuery = strtoupper($query);
    
    foreach ($forbidden as $keyword) {
        if (strpos($upperQuery, $keyword) !== false) {
            return false;
        }
    }
    
    return true;
}

function sanitize_html(string $content): string {
    $content = preg_replace('/<script[^>]*>.*?<\/script>/i', '', $content);
    $content = preg_replace('/on\w+\s*=/i', '', $content);
    return $content;
}

$result = [
    'escaped_quote' => escape_sql("user's name"),
    'valid_select' => validate_query('SELECT * FROM users'),
    'invalid_drop' => validate_query('DROP TABLE users'),
    'invalid_delete' => validate_query('DELETE FROM data'),
    'sanitized_html' => sanitize_html('hello<script>alert(1)</script>'),
    'injection_prevention_active' => true,
];

echo json_encode($result);
?>

File: mirror/tier2-dart/showcase/sec_injection.dart
// MIRROR V3: Tier 2 Dart Showcase
// Category: Security
// Module: sec_injection.dart

String escapeSql(String value) {
  return value
      .replaceAll("'", "''")
      .replaceAll('"', '""')
      .replaceAll('\\', '\\\\');
}

bool validateQuery(String query) {
  final forbidden = ['DROP', 'DELETE', 'UPDATE', 'INSERT', 'ALTER'];
  final upperQuery = query.toUpperCase();
  
  for (final keyword in forbidden) {
    if (upperQuery.contains(keyword)) {
      return false;
    }
  }
  
  return true;
}

String sanitizeHtml(String content) {
  var result = content;
  result = result.replaceAll(RegExp(r'<script[^>]*>.*?</script>', caseSensitive: false), '');
  result = result.replaceAll(RegExp(r'on\w+\s*=', caseSensitive: false), '');
  return result;
}

void main() {
  final result = <String, dynamic>{
    'escapedQuote': escapeSql("user's name"),
    'validSelect': validateQuery('SELECT * FROM users'),
    'invalidDrop': validateQuery('DROP TABLE users'),
    'invalidDelete': validateQuery('DELETE FROM data'),
    'sanitizedHtml': sanitizeHtml('hello<script>alert(1)</script>'),
    'injectionPreventionActive': true,
  };
  print(result);
}
